src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -169,9 +169,9 @@ private PerAnalyzerState GetAnalyzerState(DiagnosticAnalyzer analyzer)
         {
             var model = getCachedSemanticModel(tree, compilation, cancellationToken);
             var fullSpan = tree.GetRoot(cancellationToken).FullSpan;
-            var declarationInfos = new List<DeclarationInfo>();
+            var declarationInfos = ArrayBuilder<DeclarationInfo>.GetInstance();
             model.ComputeDeclarationsInSpan(fullSpan, getSymbol: true, builder: declarationInfos, cancellationToken: cancellationToken);
-            return declarationInfos.Select(declInfo => declInfo.DeclaredSymbol).Distinct().WhereNotNull();
+            return declarationInfos.ToArrayAndFree().Select(declInfo => declInfo.DeclaredSymbol).Distinct().WhereNotNull();
```

Comment:
>Select(declInfo => declInfo.DeclaredSymbol).Distinct().WhereNotNull() [](start = 53, length = 69)

Is this an intent to keep all these operations (select, distinct, where) lazy and repeat them for each individual enumeration through the result? If not, then consider saving the result ToArray rather than saving the source into an array, and query of off the builder directly.